### PR TITLE
Example for image classification using efficientnet with fine-tuning

### PR DESCRIPTION
**Pull request by yixingfu**

### CODE REVIEW

**Comment by qlzh727** [comment]
FYI, @fchollet and @omalleyt12

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
+"""
+Title: Image classification using EfficientNet and fine-tuning
+Author: Yixing Fu
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:3]: You can provide a link to your profile if you prefer, eg github account, etc.

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+## Compound scaling
+
+The EfficientNet models are approximately created using compound scaling. Starting from
+the base model B0, as model size scales from B0 to B7, the extra computational resource
+is proportioned into width, depth and resolution of the model by requiring each of the
+three dimensions to grow at the same power of a set of fixed ratios.
+
+However, it must be noted that the ratios are not taken accurately. A few points need to
+be taken into account:
+Resolution. Resolutions not divisible by 8, 16, etc. cause zero-padding near boundaries
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:38]: Probably add a line break or 1/2/3 here, to make individual points more readable.

Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:38]: Change to "Resolution: " and same for "Depth and width: "

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+
+An implementation of EfficientNet B0 to B7 has been shipped with tf.keras since TF2.3. To
+use EfficientNetB0 for classifying 1000 classes of images from imagenet, run
+```
+from tensorflow.keras.applications import EfficientNetB0
+model = EfficientNetB0(weights='imagenet')
+```
+
+This model takes input images of shape (224, 224, 3), and the input data should range
+[0,255]. Resizing and normalization are included as part of the model.
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:64]: space after "," for [0, 255].

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+
+batch_size = 64
+
+(x_train, y_train), (x_test, y_test) = cifar100.load_data()
+NUM_CLASSES = 100
+
+x_train = tf.cast(x_train, tf.int32)
+x_test = tf.cast(x_test, tf.int32)
+
+truncate_data = False  # @param {type: "boolean"}
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:157]: Just curious why we need this option, is it for testing with smaller amount of data? Probably add a comment for this.

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+)
+from tensorflow.keras.optimizers import SGD
+
+with strategy.scope():
+    inputs = keras.layers.Input(shape=(32, 32, 3))
+    x = inputs
+
+    x = RandomFlip()(x)
+    x = RandomContrast(0.1)(x)
+    # x = RandomHeight(0.1)(x)
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:199]: Any reason this is commented out?

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+    x = RandomFlip()(x)
+    x = RandomContrast(0.1)(x)
+    x = Resizing(IMG_SIZE, IMG_SIZE, interpolation="bilinear")(x)
+    # other preprocessing layers can be used similar to Resizing and RandomRotation
+
+    model = EfficientNetB0(include_top=False, input_tensor=x, weights="imagenet")
+
+    # freeze the pretrained weights
+    for l in model.layers:
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:273]: I think you can set model.trainable = False, which should set all the sub layer's trainable to False. Can u try that?

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    x = keras.layers.BatchNormalization()(x)
+
+    top_dropout_rate = 0.2
+    x = keras.layers.Dropout(top_dropout_rate, name="top_dropout")(x)
+    x = keras.layers.Dense(100, activation="softmax", name="pred")(x)
+
+    # compile
+    model = keras.Model(inputs, x, name="EfficientNet")
+    sgd = SGD(learning_rate=0.2, momentum=0.1, nesterov=True)
+    # sgd = tfa.optimizers.MovingAverage(sgd)
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:287]: Please remove it if it is not used.

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+    model.compile(optimizer=sgd, loss="categorical_crossentropy", metrics=["accuracy"])
+    return model
+
+
+"""
+Note that it is also possible to freeze pre-trained part entirely by
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:293]: Hum, then why didn't we do it?

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+    monitor="val_loss", factor=0.2, patience=5, min_lr=0.00001, verbose=2
+)
+epochs = 25  # @param {type: "slider", min:8, max:80}
+hist = model.fit(
+    ds_train, epochs=epochs, validation_data=ds_test, callbacks=[reduce_lr], verbose=2,
+)
+plot_hist(hist)
+
+"""
+### tips for fine tuning EfficientNet
```
Comment (qlzh727) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:364]: Tips.

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 5 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:199]: I generated output using TPU, and this layer is not supported for now.  Maybe I should generate the example using GPU instead? Also this augmentation does not change result significantly.

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 6 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:273]: Sure. As explicitly going through each layer is demonstrated when unfreezing the model, it is certainly good to use model.trainable

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 2 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:38]: added.

Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:38]: changed.

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 4 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:157]: Yes this is for testing with smaller amount of data.  The comment makes an interactive button when opened with colab. 
Sure, I should add a comment to clearly explain.

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+Title: Image classification using EfficientNet and fine-tuning
+Author: Yixing Fu
+Date created: 2020/06/30
+Last modified: 2020/07/06
+Description: Use EfficientNet with weights pre-trained on imagenet for CIFAR-100 classification.
+"""
+"""
+
+## What is EfficientNet
+EfficientNet, first introduced in https://arxiv.org/abs/1905.11946 is among the most
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:11]: Please use markdown formatting for links, e.g. 

`[Tan and Le, 2019](https://arxiv.org/abs/1905.11946)`

Applicable to all links in this file.

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+efficient models (i.e. requiring least FLOPS for inference) that reaches SOTA in both
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:12]: ...State-of-the-Art accuracy on both ImageNet...

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+
+Note: to better see validation peeling off from training accuracy, run ~20 epochs.
+"""
+
+from tensorflow.keras.applications import EfficientNetB0
+from tensorflow.keras.layers.experimental.preprocessing import (
+    Resizing,
+    RandomFlip,
+    RandomContrast,
+    # RandomHeight,
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:189]: Removed commented-out line

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 5 above)
Comment (fchollet) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:199]: Prefer RandomTranslation and RandomRotation for augmentation. More useful than RandomHeight

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 13]
```diff
@@ -0,0 +1,423 @@
...
+
+    sgd = SGD(learning_rate=0.2, momentum=0.1, nesterov=True)
+    model.compile(optimizer=sgd, loss="categorical_crossentropy", metrics=["accuracy"])
+
+model.summary()
+reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(
+    monitor="val_loss", factor=0.2, patience=5, min_lr=0.005, verbose=2
+)
+
+epochs = 20  # @param {type: "slider", min:5, max:50}
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:214]: ?

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+
+"""
+### training from scratch
+To build model that use EfficientNetB0 with 100 classes that is initiated from scratch:
+
+Note: to better see validation peeling off from training accuracy, run ~20 epochs.
+"""
+
+from tensorflow.keras.applications import EfficientNetB0
+from tensorflow.keras.layers.experimental.preprocessing import (
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:185]: Prefer importing `preprocessing` then accessing individual layers from there

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+
+## Example: EfficientNetB0 for CIFAR-100.
+
+As an architecture, EfficientNet is capable of a wide range of image classification
+tasks. For example, we will show using pre-trained EfficientNetB0 on CIFAR-100. For
+EfficientNetB0, image size is 224.
+"""
+
+# IMG_SIZE is determined by EfficientNet model choice
+IMG_SIZE = 224
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:110]: It seems very strange to resize 32x32 images to 224x244. Perhaps you should prefer a data of more naturally-sized images? (see list of TFDS supported datasets)

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+
+"""
+### prepare
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:113]: Fix title (use the `add_example` command to preview your rendered file & notebook)

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+"""
+
+"""shell
+!pip install --quiet tensorflow==2.3.0rc0
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:117]: In shell cells, don't use `!`, these are added automatically.

Rather than installing a specific TF version (which will soon no longer be necessary and which may cause bad side effects when run locally) prefer saying "This example requires TensorFlow 2.3 or above and cloud-tpu-client"

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 18]
```diff
@@ -0,0 +1,423 @@
...
+    RandomContrast,
+    # RandomHeight,
+)
+from tensorflow.keras.optimizers import SGD
+
+with strategy.scope():
+    inputs = keras.layers.Input(shape=(32, 32, 3))
+    x = inputs
+
+    x = RandomFlip()(x)
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:197]: Please factor out the data augmentation in a separate model and use that model to generate a preview of the augmented data. Visualizing what you are training on is important

See https://keras.io/examples/vision/image_classification_from_scratch/

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+    plt.ylabel("accuracy")
+    plt.xlabel("epoch")
+    plt.legend(["train", "validation"], loc="upper left")
+    plt.show()
+
+
+plot_hist(hist)
+
+"""
+### transfer learning from pretrained weight
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:248]: Please capitalize all titles

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 20]
```diff
@@ -0,0 +1,423 @@
...
+    x = Resizing(IMG_SIZE, IMG_SIZE, interpolation="bilinear")(x)
+    # other preprocessing layers can be used similar to Resizing and RandomRotation
+
+    model = EfficientNetB0(include_top=False, input_tensor=x, weights="imagenet")
+
+    # freeze the pretrained weights
+    for l in model.layers:
+        l.trainable = False
+
+    # rebuild top
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:276]: Please capitalize all code comments

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 21]
```diff
@@ -0,0 +1,423 @@
...
+hist = model.fit(
+    ds_train, epochs=epochs, validation_data=ds_test, callbacks=[reduce_lr], verbose=2,
+)
+plot_hist(hist)
+
+"""
+### tips for fine tuning EfficientNet
+
+On unfreezing layers:
+- The batch normalization layers need to be kept untrainable
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:367]: untrainable -> frozen

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 22]
```diff
@@ -0,0 +1,423 @@
...
+first epoch after unfreezing will significantly reduce accuracy.
+- In some cases it may be beneficial to open up only a portion of layers instead of
+unfreezing all. This will make fine tuning much faster when going to larger models like
+B7.
+- Each block needs to be all turned on or off. This is because the architecture includes
+a shortcut from the first layer to the last layer for each block. Not respecting blocks
+also significantly harms the final performance.
+
+
+Some other tips for utilizing EfficientNet
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:378]: Make sure to have a blank line before a list to ensure proper rendering of the list

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 23]
```diff
@@ -0,0 +1,423 @@
...
+## Using the latest EfficientNet weights
+
+Since the initial paper, the EfficientNet has been improved by various methods for data
+preprocessing and for using unlabelled data to enhance learning results. These
+improvements are relatively hard and computationally costly to reproduce, and require
+extra code; but the weights are readily available in the form of TF checkpoint files. The
+model architecture has not changed, so loading the improved checkpoints is possible.
+
+To use a checkpoint provided at
+(https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet), first
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:401]: Fix link formatting

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 24]
```diff
@@ -0,0 +1,423 @@
...
+with strategy.scope():
+    inputs = keras.layers.Input(shape=(32, 32, 3))
+    x = inputs
+
+    x = RandomFlip()(x)
+    x = RandomContrast(0.1)(x)
+    # x = RandomHeight(0.1)(x)
+    x = Resizing(IMG_SIZE, IMG_SIZE, interpolation="bilinear")(x)
+
+    x = EfficientNetB0(include_top=True, weights=None, classes=100)(x)
```
Comment (lc0) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:202]: Since you are using a constant for number of classes above
```suggestion
    x = EfficientNetB0(include_top=True, weights=None, classes=NUM_CLASSES)(x)
```

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 14 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:185]: Thanks for the suggestion. Change made.

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 15 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:110]: Yes this is strange, and there has been a lot of questions around people trying to reproduce the EfficientNet result as they claim very high CIFAR100 transfer learning. Upsampling 32x32 image is also what the authors claimed what they did. 

Picking a higher resolution dataset should (not tried yet though) actually make the result better. Let me try a few.

File: examples/vision/image_classification_efficientnet_fine_tuning.py (hunk 13 above)
Comment (yixingfu) [reply examples/vision/image_classification_efficientnet_fine_tuning.py:214]: I created the example from colab, and this comment adds an interactive toggle when opened there. Should I remove this if it is confusing? @fchollet

File: examples/vision/image_classification_efficientnet_fine_tuning.py [hunk 25]
```diff
@@ -0,0 +1,414 @@
...
+
+Note: to better see validation peeling off from training accuracy, run ~20 epochs.
+"""
+
+from tensorflow.keras.applications import EfficientNetB0
+from tensorflow.keras.layers.experimental.preprocessing import (
+    Resizing,
+    RandomFlip,
+    RandomContrast,
+)
```
Comment (fchollet) [inline examples/vision/image_classification_efficientnet_fine_tuning.py:196]: from tensorflow.keras.layers.experimental import preprocessing

**Comment by irvifa** [comment]
Please rebase from master, there's a fix for the CI

**Comment by MarkDaoust** [comment]
This looks like something messed up in a merge commit. Can you push a clean version of the commit history or create a new pr, rebased on  master?

**Comment by yixingfu** [comment]
Sure. The merge after rebase seemed to mess it up
